### PR TITLE
fix: added a call limit setting to prevent crashes or timeouts

### LINKED ISSUE #674: oss-fuzz: stack overflow in `pest::parser_state::ParserState<R>::rule`
**Describe the bug**

Fuzzing found a stack overflow. Detailed report: https://oss-fuzz.com/testcase-detail/4831410241798144

Upstream tracker: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=49508

**To Reproduce**

https://google.github.io/oss-fuzz/advanced-topics/reproducing/

**Expected behavior**

No stack overflow.

**Additional context**

crash revision: e30f63e9a5b0967a2d5a2f6ef7b7d8edfc526d3f

🙃 I'm getting cc'd at my primary email and can access the oss-fuzz.com details but not the bugs.chromium.org tracker. (https://github.com/google/oss-fuzz/issues/8096)

### LINKED ISSUE #675: oss-fuzz: timout in parser
**Describe the bug**

Fuzzing found a timeout. Detailed report: https://oss-fuzz.com/testcase?key=6048143472459776

Upstream tracker: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=49513

**To Reproduce**

https://google.github.io/oss-fuzz/advanced-topics/reproducing/

**Expected behavior**

Completing within the timeout window (60s).

**Additional context**

crash revision: e30f63e9a5b0967a2d5a2f6ef7b7d8edfc526d3f

🙃 I'm getting cc'd at my primary email and can access the oss-fuzz.com details but not the bugs.chromium.org tracker. (https://github.com/google/oss-fuzz/issues/8096)

### PR DESCRIPTION
**Pull request by tomtau**
closes #674, closes #675

### CODE REVIEW

File: pest/src/parser_state.rs [hunk 1]
```diff
@@ -50,6 +52,63 @@ pub enum MatchDir {
...
+/// to prevent stack overflows or excessive execution times
+/// in some grammars.
+///
+/// # Arguments
+///
+/// * `limit` - The maximum call depth or the number of calls.
+///             If 0, the call depth or the number of calls is unlimited.
+/// * `finite_calls` - If `true`, the parser execution is limited to
+///                    a finite number of calls (based on `limit`).
+pub fn set_call_limit(limit: usize, finite_calls: bool) {
```
Comment (glyn) [inline pest/src/parser_state.rs:68]: Wouldn't the API be more natural if it took a single argument of type `Option<usize>`?

Comment (tomtau) [reply pest/src/parser_state.rs:68]: Possibly, what would be the difference between calling it with `None` vs `Some(0)`?

Comment (tomtau) [reply pest/src/parser_state.rs:68]: If you say a single argument, maybe the description isn't clear. For completeness, the second argument is the following:
* `true` -- it won't be decrementing the counter when exiting functions, i.e. it'll terminate after the set number of calls
(this is against timeouts when parsing potentially adversarial inputs on some grammars)
* `false` -- it'll continue executing as long as the call depth stays within the set limit
(this is if one wants to make sure it doesn't crash with "stack overflow" when parsing some inputs on some grammars)

Comment (tomtau) [reply pest/src/parser_state.rs:68]: Maybe in practice, one will only/mostly want the former option (`true`)?

Comment (glyn) [reply pest/src/parser_state.rs:68]: > Possibly, what would be the difference between calling it with `None` vs `Some(0)`?

`Some(0)` could be construed to mean the same as `None`, i.e. unlimited call depth. But in that case, it might be better to have a single argument of type `usize` with a value of `0` meaning unlimited call depth.

Or we could avoid attaching a special meaning to `0` and have the implementation hit the call limit immediately, i.e. before any calls.

Comment (glyn) [reply pest/src/parser_state.rs:68]: > Maybe in practice, one will only/mostly want the former option (`true`)?

Perhaps. I guess anyone wanting an unlimited call depth could use a limit of `usize::MAX`, although that would be pretty ugly as an API call.

Comment (glyn) [reply pest/src/parser_state.rs:68]: On balance, I thing a single argument `limit: usize` with the value of `0` meaning unlimited is pretty clean, and certainly cleaner than having a `usize` and a `bool`.

Comment (NoahTheDuke) [reply pest/src/parser_state.rs:68]: Maybe this is my limited Java experience showing through but I would expect these to be two separate functions. I agree that it feels odd to have the bool included here, at least. Maybe the finite call fn could be `set_finite_calls() { FINITE_CALL_LIMIT.store(true, Ordering::Relaxed); }` or something similar?

I don't know that I agree that using an `Option` is helpful unless we're going to disallow `0` as a valid input: `None` or `Some(100)` but never `Some(0)`.

Comment (dragostis) [reply pest/src/parser_state.rs:68]: Why not use `Option<NonZeroUsize>`?

**Comment by dragostis** [comment]
#675 looks like it shouldn't be closed by this since it's a regression. Ideally, the optimizer should be able to detect all exponential cases and have replacement rules for them. Maybe someone could take a look at [egg](https://egraphs-good.github.io/) and implement a simplified version of the grammar there so peephole optimizations can be more easily written?

File: pest/src/parser_state.rs (hunk 1 above)
Comment (tomtau) [reply pest/src/parser_state.rs:68]: > Perhaps. I guess anyone wanting an unlimited call depth could use a limit of usize::MAX, although that would be pretty ugly as an API call.

usize::MAX could still cause a stack overflow in that situation; but yes, depending on the grammar, one could just set a reasonable limit of finite calls that would parse common cases and wouldn't crash on adversarial inputs

> Why not use `Option<NonZeroUsize>`?

That's a good suggestion, I think we can go for it. The downside is a bit wordy calling of that function with: `set_call_limit(Some(xxx.try_into().unwrap()),...)` instead of `set_call_limit(xxx, ...)`

**Comment by tomtau** [comment]
@dragostis thanks. Yes, I agree that for the timeout issue, there should be a fix for the root cause. For the edge case found by the fuzzer, it seems it's due to the block comment feature introduced in this PR: https://github.com/pest-parser/pest/pull/332
There are, however, a few other open issues related to performance which may or may not have the same root cause, e.g. https://github.com/pest-parser/pest/issues/603 https://github.com/pest-parser/pest/issues/571 https://github.com/pest-parser/pest/issues/402 https://github.com/pest-parser/pest/issues/279

Given the investigation and the corresponding proper backwards-compatible fix for timeouts may take time, there are potentially two ways (with regards to this API: https://github.com/pest-parser/pest/pull/684#discussion_r941397159 ) to keep `set_call_limit` with a single argument:

1. make it track overall calls (`true`). This should fix those fuzzer issues, as demonstrated by the unit tests.
2. make it track call depths (`false`). This should fix only the stack overflow crashes, but not the timeout issues.

I'm leaning towards the first one, because, in the meantime, it's practical in the sense it'd allow people to use pest on potentially problematic inputs (without worrying it'd eat up all CPU time) if their grammars suffered from this issue.
I'd then open up an issue that would reference the fuzzer timeout issue as well the other performance issues.

--
BTW, as for the solution, one potential option that comes to mind is packrat parsing with a fix: https://dl.acm.org/doi/10.1145/3377555.3377898 There's one older issue that was closed https://github.com/pest-parser/pest/issues/7#issuecomment-323311318 -- do you have any more insight into why packrat parsing was not used?

**Comment by dragostis** [comment]
Packrat parsing was not used because caching data is slow. While it removes the exponential cases, it adds substantial overhead for all other parsing.

My solution to this was to identify such patterns and replace them in the optimizer. This has worked for some cases, but I haven't had the time to identify all patterns of interest. Also, I have no proof that all such patterns are actually replaceable with equivalent but non-exponential patterns.

My vision for pest when I started it was to actually use something like [ruler](https://github.com/uwplse/ruler) to find these rules and use egg to optimize the grammars. Currently, I think this is a bit hard since pest's grammar is in need of a cleanup. I started working on a new meta grammar in [pest3](https://github.com/pest-parser/pest3) a few years ago, but didn't have a chance to make it work. It has a few interesting ideas, including much better code generation, if you have the time to take over that work.

**Comment by tomtau** [comment]
@glyn The API was simplified to only track calls as per the discussion: https://github.com/pest-parser/pest/pull/684#discussion_r941383935

`Option<NonZeroUsize>` makes it a bit wordy with literals, but that may be improved in future Rust versions:
https://github.com/rust-lang/rust/issues/69329

File: pest/src/parser_state.rs [hunk 2]
```diff
@@ -50,6 +53,48 @@ pub enum MatchDir {
...
+/// Sets the maximum call limit for the parser
+/// to prevent stack overflows or excessive execution times
+/// in some grammars.
+///
+/// # Arguments
+///
+/// * `limit` - The maximum call depth or the number of calls.
+///             If 0, the call depth or the number of calls is unlimited.
+/// * `finite_calls` - If `true`, the parser execution is limited to
+///                    a finite number of calls (based on `limit`).
```
Comment (glyn) [inline pest/src/parser_state.rs:67]: These docs don't match the API.

**Review by glyn — CHANGES_REQUESTED** [review]
Please update the docs of `set_call_limit`.

File: pest/src/parser_state.rs (hunk 2 above)
Comment (tomtau) [reply pest/src/parser_state.rs:67]: oh, thanks!

File: pest/src/parser_state.rs [hunk 3]
```diff
@@ -61,10 +61,8 @@ static CALL_LIMIT: AtomicUsize = AtomicUsize::new(0);
 ///
 /// # Arguments
 ///
-/// * `limit` - The maximum call depth or the number of calls.
-///             If 0, the call depth or the number of calls is unlimited.
-/// * `finite_calls` - If `true`, the parser execution is limited to
-///                    a finite number of calls (based on `limit`).
+/// * `limit` - The maximum number of calls. If None,
+///             the number of calls is unlimited.
```
Comment (glyn) [inline pest/src/parser_state.rs:65]: That's better thanks. I wonder if you could clarify in the doc comment whether "number of calls" means the same as call depth or is just a running total. Also, please define what constitutes a "call"?

**Review by glyn — APPROVED** [review]
Remove blocking review. Please note I have only reviewed a small part of this PR.

File: pest/src/parser_state.rs (hunk 3 above)
Comment (tomtau) [reply pest/src/parser_state.rs:68]: After https://github.com/pest-parser/pest/pull/684#issuecomment-1210456883 , it's a running total. I've expanded the doc comment.

**Review by glyn — APPROVED** [review]
LGTM, thanks.

File: pest/src/parser_state.rs [hunk 4]
```diff
@@ -50,6 +53,49 @@ pub enum MatchDir {
...
+impl CallLimitTracker {
+    fn limit_reached(&self) -> bool {
+        self.current_call_limit
+            .map_or(false, |(current, limit)| current >= limit)
+    }
+
+    fn increment_depth(&mut self) {
+        if let Some((current, limit)) = self.current_call_limit {
+            self.current_call_limit = Some((current + 1, limit));
+        }
```
Comment (NoahTheDuke) [inline pest/src/parser_state.rs:95]:
```suggestion
        if let Some((current, _)) = &mut self.current_call_limit {
            *current += 1;
        }
```

**Review by NoahTheDuke — APPROVED** [review]
Left a suggestion but take it or not, this is a good change.

File: pest/src/parser_state.rs [hunk 5]
```diff
@@ -50,6 +53,49 @@ pub enum MatchDir {
...
+impl CallLimitTracker {
+    fn limit_reached(&self) -> bool {
+        self.current_call_limit
+            .map_or(false, |(current, limit)| current >= limit)
+    }
+
+    fn increment_depth(&mut self) {
+        if let Some((current, _)) = &mut self.current_call_limit {
+            *current += 1;
+        }
```
Comment (glyn) [inline pest/src/parser_state.rs:95]: @tomtau @NoahTheDuke 
Just for interest, there's yet another way to code this:
```
        if let Some((ref mut current, _)) = self.current_call_limit {
            *current += 1;
        }
```
This cropped up [here](https://youtu.be/rAl-9HwD858?t=2752) in @jonhoo's video [Crust of Rust: Lifetime Annotations](https://www.youtube.com/watch?v=rAl-9HwD858).

Comment (NoahTheDuke) [reply pest/src/parser_state.rs:95]: Oh that's cool. I didn't know you could embed those.
